### PR TITLE
[Build] Move to npm-based build

### PR DESCRIPTION
**Pull request by VWoeltjen**
Opening PR for #460 early, for visibility and progress tracking.

Build steps to add (from #459, mostly) and other tasks:

- [x] `karma` to run tests (unit tests and functional tests)
- [x] `istanbul` (via `karma-coverage`) to capture code coverage information.
- [x] `r.js` (RequireJS optimizer) to minify JavaScript sources into a single artifact (or small number of artifacts)
- [x] `compass` (via `gulp-compass`) to compile SASS->CSS and output within an `assets` hierarchy
- [x] `text` plugin for RequireJS to compile templates into a single artifact
- [x] Generate docs (~~#225 to produce documentation (both JSDoc, and from Markdown sources)~~ is deferred to #225)
- [x] `jshint` for static analysis / code correctness
- [x] `jscs` for code style
- [x] Move `index.html` into `dist`
- [x] Generate build info for About dialog (timestamp, commit)
- [x] Remove `pom.xml`
- [x] Update build-related documentation
- [x] ~~Handle asset paths (likely involves moving bundles into `src` hierarchy, removing `res` directories)~~ Accounted for in followup #650
- [x] Update tests for changed files (may want to just `xit` failing tests in cases where code is expected to become obsolete soon)
- [x] Update JSDoc for changed files
- [x] Trigger gulp tasks from appropriate npm build steps
- [x] Allow removal of registered bundles (e.g. `legacyRegistry.remove()`)
- [x] Update licensing info

### CODE REVIEW

**Comment by VWoeltjen** [comment]
@larkin / @akhenry - want to run some questions about scope by you.

One of the points we agreed to in the context of #69 was to switch to a directory hierarchy which placed related scripts/tests/mocks/templates/etc as peers, e.g.:

```
+ src
|
+-+ foo
  |
  + FooController.js
  |
  + FooControllerMock.js
  |
  + FooControllerSpec.js
  |
  + foo.html
  |
  + foo.scss
```

In other words, get rid of the `res` / `test` / `src` directories entirely.

I'm thinking it makes sense to do this reorganization in the context of this issue. Particularly, I'm looking at some of the changes itemized above (`compass` compilation; usage of the `requirejs-text` plugin to allow templates to be compiled-in) and thinking it makes sense to do these against the directory structure we actually want, instead of the structure we have now.

On the other hand, I'd like to handle this issue in the smallest chunks possible, to minimize merge conflicts, allow better prioritization, etc...

Thoughts? Do you see a better way to order the work associated with #460 (particularly one that would create fewer conflicts later)?

**Comment by VWoeltjen** [comment]
Think this is ready to merge. Would also like to file follow-up issues for:

* `karma-coverage` configuration (doubles as a way of reviewing/formalizing code coverage requirements)
* `jshint` configuration
* `jscs` configuration (doubles as a way of discussing code style standards)
* `text` plugin utilization (such that HTML templates get compiled-in; should occur after or during #650)

Have not filed these yet in case we want to do some portion of this in the context of this PR.

### Author Checklist

1. Changes address original issue? Y
2. Unit tests included and/or updated with changes? Y
3. Command line build passes? Y
4. Changes have been smoke-tested? Y

File: platform/representation/src/TemplateLinker.js [hunk 1]
```diff
@@ -84,10 +84,11 @@ define(
...
          * @param element the jqLite-wrapped element into which templates
          *        should be inserted
-         * @returns {Function} a function which can be called with a template
-         *          URL to switch templates, or `undefined` to remove.
+         * @returns {Function} a function which can be called with a template's
+         *          extension definition to switch templates, or `undefined`
+         *          to remove.
          */
-        TemplateLinker.prototype.link = function (scope, element, templateUrl) {
+        TemplateLinker.prototype.link = function (scope, element, ext) {
```
Comment (larkin) [inline platform/representation/src/TemplateLinker.js]: Need a docstring for `ext` param.

**Comment by larkin** [comment]
There's a small JSDoc issue but this otherwise looks good.  Going to send back to @VWoeltjen to fix the docstring and merge.

# Reviewer Checklist

1. Changes appear to address issue? Y
2. Appropriate unit tests included? Y
3. Code style and in-line documentation are appropriate? Y
4. Commit messages meet standards? Y

**Comment by VWoeltjen** [comment]
Addressed JSDoc issue (good catch!), completing merge now.

**Pull request merged**